damnit/backend/db_migrations.py

```diff
@@ -0,0 +1,181 @@
+import shutil
+from dataclasses import dataclass
+from datetime import datetime, timezone
+from pathlib import Path
+from textwrap import dedent
+from typing import Callable, Iterable
+
+import sqlite3
+
+
+@dataclass(frozen=True)
+class Migration:
+    """
+    A single schema migration step applied to go to `to_version`.
+
+    - `to_version`: schema version after this migration completes
+    - `description`: short human description for docs/logs
+    - `apply`: a callable that takes a sqlite3.Connection and applies changes
+    """
+
+    to_version: int
+    description: str
+    apply: Callable[[sqlite3.Connection], None]
+
+
+def _to_v1(conn: sqlite3.Connection) -> None:
```

Comment:
Purely stylistic, but it would be neat to associate the numbers and the descriptions more closely with the functions. Maybe via a decorator:

```python
MIGRATIONS = []

def migration_func(to_version):
    def inner(func):
        MIGRATIONS.append(Migration(to_version, func.__doc__, func))
    return inner

@migration_func(1)
def _to_v1(conn: sqlite3.Connection):
    """Baseline schema (core tables, views)"""
    ...
```